WIP: Improve url previews in Zulip

punchagan opened the pull request:
<!-- What's this PR for?  (Just a link to an issue is fine.) -->

Trying to address #9451

punchagan commented:
The current implementation is not using `oembed` the way I think it is expected to be used.  `oembed` responses seem to have `html` that can be embedded into our rendered message embed. But, we have only been using it to get `title`, `description` and `image`. In some cases (spotify, for example), the `oembed` response is worse than what we get from the open graph tags in the html. 

![image](https://user-images.githubusercontent.com/315678/57118029-0c3ad580-6d7e-11e9-9bd7-36704e7f432c.png)

I've tried to hack something up to try and use the `html` in the `oembed` response, and it looks something like this for spotify. 
![image](https://user-images.githubusercontent.com/315678/57117979-a6e6e480-6d7d-11e9-84a7-3e4894dd64a2.png)

@timabbott what are your thoughts on trying to use the `html` from the oembed responses? 

If we are not using it, I'd vote for just doing simple open-graph headers parsing and using that. It will at least make the code easier to understand, and remove an external dependency (which isn't being very actively maintained).

punchagan commented:
![image](https://user-images.githubusercontent.com/315678/57118802-4d81b400-6d83-11e9-9227-2d36d87448d2.png)

Modified the CSS a little to make it look better

timabbott commented on zerver/lib/url_preview/preview.py:
```diff
@@ -10,7 +11,7 @@
 
 
-CACHE_NAME = "database"
+CACHE_NAME = "database" if not settings.DEVELOPMENT else "in-memory"
```
What's the story with this change?

punchagan replied:
It's difficult to test changes to the code in development, because of the caching. Using an in-memory cache makes this easier, since each code change flushes the cache. (This was also listed as a potential option in the original issue #9451)

punchagan replied:
I guess the commit message could use an improvement, if we agree this change is acceptable.

timabbott commented:
I would be very happy to get rid of the external dependency (or look for a better one).  A few random notes on issues:
* Our current setup doesn't use HTTP HEAD requests, and thus if you paste a giant file that it thinks is a previewable page, it'll download it.  Has been a cause of OOM kills in the past.
* It's sorta hard to reason above.
* For open graph HTML, who is generating that HTML?  We definitely want to be thoughtful about XSS issues for content coming from these.

punchagan commented:
> For open graph HTML, who is generating that HTML? We definitely want to be thoughtful about XSS issues for content coming from these.

The oEmbed HTML is provided by an [oembed service](https://oembed.com/#section1) run by the website whose preview we are fetching. And we definitely need to be [thoughtful about XSS](https://oembed.com/#section3) and may be even have a list of approved providers, etc. The HTML comes often comes with custom Javascript from the service, which allows interacting with the service from within Zulip. Spotify's embed HTML lets you play music from inside Zulip, for instance. 
Slack has similar support for some websites.  But, I guess we can do this in a second stage. 

The HTML previews using open graph metadata is generated by us. We use the `title`, `description` and `image` metadata to do this. As a first step, it might be worth getting this to work well, especially things like avoiding downloading files, etc.

punchagan commented:
I've modified the code to try and use the oembed HTML to use it only if it looks safe. This will need some more careful thought and research, but may be a good start.

punchagan commented:
@timabbott I've improved this PR, based on your inputs. It would be very helpful, if you could take a 
look at it one more time. :)

Here's a brief summary of the changes in this PR. 

- Switch to using in-memory caching for the development environment, to make it easy to work on this code. 
- Refactor some test code to make it easy to create mock request objects
- Added a new dependency on `python-magic`, a wrapper around `libmagic` to sniff content-type based on content. This is useful when the server doesn't provide a `content-type` header, and we want to decide whether or not to process the URL further, to generate previews. 
- Bail out early from the code to generate previews for URLs if the content type is not `text/html`. We look at the server header, or try to guess it when the server doesn't provide one. 
- Use the `oembed` metadata's `html` directly to show "more interactive previews", when the URL in question has a valid `oembed` provider and data. To make sure the `html` is safe to use, we only accept `html` which is in an `iframe` that comes from a `http(s):` source, and not `javascript:` source. 
- As an experiment, remove our custom code for Vimeo previews, and use the `oembed` previews. 

A few questions I had are noted as FIXMEs in the code. 
- Should we be using the database cache for these previews? Changes to the metadata of a site will not reflect until the cache becomes full, and the flush clears the cached value for a URL. 
- If trying to fetch a site fails (due to a network error) or something, should we cache that value? This may not be an important question if the cache can be refreshed relatively easily, but otherwise can be a tricky thing. 
- Should we try to guess the content-type of a URL ourselves without even asking the server. Should we trust what the server tells us?

punchagan commented:
I pushed a few more commits on this PR. 

- Pushed a couple of bugs with the generic parser and the open graph parser
- Use a custom user agent when trying to fetch URLs, to prevent getting blocked by some sites. 
- Added a hackish tool to send separate lines listed in a file, as messages. This can be used to test open graph previews by using a file with a bunch of URLs, one per line. I used this tool to test with about 300 URLs taken from Hacker News. I fixed the problems I found in this testing.

timabbott commented:
I merged the first few commits.  

> Should we be using the database cache for these previews? Changes to the metadata of a site will not reflect until the cache becomes full, and the flush clears the cached value for a URL.

I think we do want to use a database-level cache in production if we are caching at all.  Given that we render markdown once (when a message is sent), I don't think there's much value in caching here unless lots of folks are sending links to the same site in different messages.  However, in that case, it's a significant latency boost in many cases.  Let's leave it in for now and see if we get complaints.  

> If trying to fetch a site fails (due to a network error) or something, should we cache that value? This may not be an important question if the cache can be refreshed relatively easily, but otherwise can be a tricky thing.

I think caching network errors is often counterproductive, though there's certainly an argument for doing something different in this case, given that broken links are much more likely in this code path (with user-entered content) than elsewhere.  Let's start with not caching them and go from there.

> Should we try to guess the content-type of a URL ourselves without even asking the server. Should we trust what the server tells us?

Given that we only want to process text/html (unless I'm missing something), I think if we're just trying to determine if the URL is HTML, the following algorithm is reasonable:
* If the server claims a type other than text/html, ignore it.
* If the server claims it is text/html, then do the content-type guessing

timabbott commented:
I merged the first several commits, thanks @punchagan!  I stopped with the magic ones mainly because that's where the FIXMEs begin.

timabbott commented:
For `Directly embed the oEmbed html if appears safe`, are there articles/specs on how that the oEmbed HTML is supposed to work that you used to figure out how that should work?

timabbott commented on static/styles/zulip.scss:
```diff
@@ -1560,7 +1560,7 @@ a:hover code {
         border: none;
         border-left: 3px solid hsl(0, 0%, 93%);
-        height: 80px;
+        min-height: 80px;
```
So, we are generally pretty careful to set a fixed height (100px) for our inline image previews, because that's critical to the scrolling UX.  Without a fixed height for areas containing external resources, the browser cannot lay out the message list exactly as it will be after external resources load, which creates bouncing of the message list as images load.

I suspect the same issue would apply here...

punchagan replied:
Yeah, that makes sense. I'll look into the UI aspects of it, after fixing the backend parts of this PR.

punchagan replied:
I've updated this to use the height provided by the oEmbed provider for the height of the embed div.

punchagan commented:
> I merged the first few commits.

Thanks for the review and the merge, Tim!

> I think we do want to use a database-level cache in production if we are caching at all. Given that we render markdown once (when a message is sent), I don't think there's much value in caching here unless lots of folks are sending links to the same site in different messages. However, in that case, it's a significant latency boost in many cases. Let's leave it in for now and see if we get complaints.

Fair enough. There's also edits of messages and replies to messages, where the caching might come in handy. (Though in case of replies, we may want to turn off the URL previews).

> I think caching network errors is often counterproductive [..] Let's start with not caching them and go from there.

Sounds good!


> the following algorithm is reasonable:
>     * If the server claims a type other than text/html, ignore it.
>     * If the server claims it is text/html, then do the content-type guessing

Perfect! I'll change the algorithm I used to match with this.

punchagan commented:
> For `Directly embed the oEmbed html if appears safe`, are there articles/specs on how that the oEmbed HTML is supposed to work that you used to figure out how that should work?

My main reference was the spec at https://oembed.com/ (section 3, in particular). I haven't really used anything else for reference, but a quick web search pops up a few articles. Vimeo, for instance, has an [example](https://github.com/vimeo/vimeo-oembed-examples/blob/master/oembed/js-example.html) of how to use their oembed.

punchagan commented:
> a quick web search pops up a few articles

[This article](https://blog.codepen.io/documentation/api/oembed/) on how to use codepen's oembed gives a reasonable overview.

timabbott commented:
For security, the recommendation is to do an iframe hosted from another domain; does this implementation do the "another domain" part?

punchagan commented:
> For security, the recommendation is to do an iframe hosted from another domain; does this implementation do the "another domain" part?

I think it does, unless I have misunderstood the recommendation. "another domain" here is a domain different from where Zulip is hosted. So, if we ensure that only iframes are included in the html and they have `src` set to `http://` or `https://` or `//`, we should be good. If the src URL is the same domain as the zulip server, I guess we can trust the js hosted there. If it is another site, it is actually "another domain" that the spec recommends. 

Does that make sense? Or am I missing something?

punchagan commented:
> I think caching network errors is often counterproductive, though there's certainly an argument for doing something different in this case, given that broken links are much more likely in this code path (with user-entered content) than elsewhere. Let's start with not caching them and go from there.

@timabbott Do you have suggestions on how to ensure that the network errors are not cached? I'm unable to think up a solution that feels good to me.

punchagan commented:
@timabbott I've addressed everything from the last review. Please take a look when you get some time. Thanks!

timabbott commented on zerver/tests/test_link_embed.py:
```diff
@@ -374,3 +375,116 @@ def test_link_embed_data_from_cache(self) -> None:
+            with mock.patch('requests.get', mocked_response):
+                FetchLinksEmbedData().consume(event)
+
+                cached_data = link_embed_data_from_cache(url)
```
Just a note on test style for this system: Long after this test suite was originally written, we changed the default behavior of `queue_json_publish` in the test suite to just call the processor function (it reaches into `zerver/worker/queue_processors.py` to figure out which that is), which provides much better test coverage and requires a lot less messy locking.  So aside from the couple tests in this file where we're expressly trying to test the situation before the queue processor runs, we should be able to simplify these tests a lot.

Secondly, for mocking HTTP requests, there's an option to use the standard `httpretty` library rather than mocking `requests.get`; might not be right for some use cases, but I wanted to mention it since it can be handy (we use it in our authentication backend tests).

punchagan replied:
Yeah, the test suite did look more complicated than other places that I touched in Zulip, but didn't investigate why. I'll simplify the tests, at some point.

timabbott commented on zerver/tests/test_link_embed.py:
```diff
@@ -374,3 +375,116 @@ def test_link_embed_data_from_cache(self) -> None:
+        headers = {'content-type': 'application/octet-stream'}
+        mocked_response = mock.Mock(side_effect=self.create_mock_response(url, headers=headers))
+
+        with self.settings(TEST_SUITE=False, CACHES=TEST_CACHES):
```
I'd also like to get away from the pattern of `self.settings(TEST_SUITE=False)`; that's definitely a land mine waiting to blow up on us.  I don't remember why we had it.

punchagan replied:
I'll dig around and see if I can figure out why we have this pattern. :+1:

timabbott commented:
I merged several more (including the ones that added the tests I commented on above, because I think we can fix forward, especially since the model is present in the existing tests).

I didn't merge the oEmbed HTML and Vimeo changes.  I think for video players in particular (and maybe basically all embeds), we probably want the same UX model that we have for the Youtube player, where we have a smaller embed object (too small to reasonably watch a video in) and you click it to open it in the Zulip lightbox.  If I'm understanding the system correctly, it should be possible to do something like the following and get the same experience we have for YouTube today for generic oEmbed services:

* For what we display in the message feed, use the open graph data for the object + the styling for our YouTube buttons (100px height image preview etc.).
* Render a `play` button on the image if the oEmbed data suggests the thing is playable (I don't know enough to know whether there are oEmbed things where one doesn't want a play button).
* Save the `iframe` embed HTML in a `data-lightbox-html` attribute, and use that when opening the lightbox to provide the player/viewer for the third-party content.  

@rishig FYI for this discussion -- the main question for you is confirmation that we want the YouTube style of UX (with a 100px height widget that you can click on to get the lightbox-style "full" widget) for generic other providers that have a way to embed video/sound/etc via the oEmbed standard.

rishig commented:
yeah, that's a good question. Even at 200px height it becomes plausible to play "inline", which is sort of nice since you can read the discussion about it (and/or type a response) while it's playing. I think I'm tempted by the 200px inline thing, if it's not more work. 
If it more work, the current youtube solution is fine and we can revisit it later.

If we do go with the current model, we should try to make it so that it starts playing when you open the lightbox, rather than having to click a second play button.

punchagan commented:
> I merged several more (including the ones that added the tests I commented on above, because I think we can fix forward, especially since the model is present in the existing tests).

Thanks Tim! I had a few improvements to the commits you merged, but my push seems to have failed. :(
I'll add a few new commits for the improvements, mainly the fix for not caching the network errors.

punchagan commented:
It can sometimes be convenient to have the "embed" playing, without the lightbox, especially if it's an audio embed, or a song video, etc. But, it does make sense to make everything consistent. 

One of the problem with putting the embed in without using lightbox is the height of the embed. The API specifies that clients should be able to pass a maxheight and the service should respect it. But, most services don't seem to do so. For instance, the Spotify embed is always 330px high. 
![image](https://user-images.githubusercontent.com/315678/57675184-d4197980-763e-11e9-85d6-b80ebdaa66c0.png)

Regarding the play button, there are other kinds of sites which also have oEmbed services for example NYT. I'm not sure this oEmbed is much more useful from just using the open graph metadata. 

![image](https://user-images.githubusercontent.com/315678/57674828-ee068c80-763d-11e9-95b4-9bc544ebf114.png)

I think it would be better to start with a whitelist of services for which we want to use the oEmbed, and customize things for them. And then slowly expand the whitelist to include other websites.

timabbott commented:
@rishig just for some context, oEmbed is sorta designed for the Facebook use case, so if you've seen the super-giant embeds of websites on Facebook that basically make the screen unusable, that's what third-party services will do by default.  So for NYT and similar sites, we will definitely want to prefer doing open graph data, so that we can control the size and formatting to be consistent and reasonable.  That's why I was imagining we should be doing oEmbed only for video/audio players.

I don't think a non-lightbox model for playing audio/video in Zulip is going to be a super great experience in any case, because if you renarrow the embed will be removed from the HTML (effectively killing your playback), and that behavior will probably include cases where we just rerender the HTML because an adjacent new message came in that required doing so.

(And in any case we should make it convenient to e.g. middle-click to pop out the player into its own window)

Having a whitelist makes sense in the near term; that way we can test individual sites for making sense before adding them (I could also imagine a whitelist for a non-lightbox experience), but ideally we'd not require a whitelist long-term.

rishig commented:
> That's why I was imagining we should be doing oEmbed only for video/audio players.

ah yeah, definitely. I think 200px makes sense at most for when you directly link to an image or video, and everything else should be 100px. 

> because if you renarrow the embed will be removed from the HTML (effectively killing your playback)

I think this is fine, and how most products work. It's the same with images -- when you renarrow you no longer see the image.
(I guess I'm mainly thinking about video here, audio always opening in another tab seems fine.)

> and that behavior will probably include cases where we just rerender the HTML because an adjacent new message came in that required doing so.

yeah, agreed not worth solving that problem for this. 

fwiw I do think the inline experience is nice (slack has it, with 200px height, if you want to play with it), but not worth delaying the rest of this feature for that given that it'll be hard to get the rerendering cases correct.

> we should make it convenient to e.g. middle-click to pop out the player into its own window

Or have a button on hover that opens it in a new window.

punchagan commented:
The oEmbed protocol allows for 4 different types of oEmbeds - `photo`, `video`, `link` and `rich`. 

Based on the discussion above, I think we implement the exact same behavior we use for images and youtube videos for the `photo` and the `video` types, respectively. Essentially, show a small thumbnail (100px?), and clicking on that opens the lightbox with the video or the photo. 

The metadata in the `link` type is quite similar to what we get from the open graph meta tags. We can show previews similar to what we would do, if we got the data from open graph meta tags. 

The `rich` type is a tricky one. It includes websites that just show a picture and some description (like the NYT example I shared above) and also interactive embeds like the Spotify Audio player. It's not easy to distinguish between these two, and I think the initial implementation can just ignore this oEmbed type and use open graph meta tags for the preview.  Later, we could decide to do special things for sites like Spotify, Soundcloud, etc.

punchagan commented:
@rishig @timabbott Does this seem like a reasonable plan? Or do you have other suggestions for the next steps?

timabbott commented:
From a technical perspective, I think going with the approach you propose @punchagan makes a lot of sense.  (Given the rerendering issues involved here, I think we should do the simpler integration without inline playback in the message feed (which will involve figuring out a lot of the pieces involved for any solution) and then go from there.).  Let's keep it at 100px; if we want to move to 200px, we'd probably want to move the existing video/image previews too, not just oembed ones, so that's sorta an independent change.

punchagan commented:
Thanks Tim! I'll try to work on this, during this week.

timabbott commented:
I merged the caching commit; for the oEmbed one, https://oembed.com/#section3 specifically mentions an iframe hosted from another domain.  I'm 

(Also, what do you want to do about the `FIXME`?  Seems like it might be better to just not add a preview when there's no image for now?)

punchagan commented:
I may be misunderstanding the "another domain" part, but I had [posted a comment](https://github.com/zulip/zulip/pull/12250#issuecomment-489915131) a few weeks back, that you may have missed. Does it make sense?

punchagan commented:
> (Also, what do you want to do about the `FIXME`? Seems like it might be better to just not add a preview when there's no image for now?)

Yeah, that makes sense!

punchagan commented:
> I may be misunderstanding the "another domain" part, but I had [posted a comment](https://github.com/zulip/zulip/pull/12250#issuecomment-489915131) a few weeks back, that you may have missed. Does it make sense?

Also, the embeds would right now be exactly the same as YouTube or Vimeo embeds we have right now. So, I guess we should be good, though the sites we are embedding aren't from a whitelist chosen by us (currently, youtube and vimeo).

punchagan commented:
@timabbott I force pushed a couple of commits that I think are ready for another review.

timabbott commented on zerver/lib/url_preview/oembed/__init__.py:
```diff
@@ -11,4 +11,10 @@ def get_oembed_data(url: str,
         return None
 
     data['image'] = data.get('thumbnail_url')
+    type_ = data.get('type', '')
```
I would call this `open_graph_type`?

punchagan replied:
Yeah, It's definitely a better name.  I'll send in a refactor commit when I'm fixing up the vimeo previews stuff. Thanks!

timabbott commented:
I merged the first two commits.  For the vimeo one, we can't remove the frontend/lightbox code, because it will break previews of previously rendered messages.  So I think we need to make that frontend code just have comments specifying that we no longer generate that and it's there for legacy messages rendered before we migrated to a using generic oembed to manage vimeo, with a link to this comment on GitHub.

(Might be able to remove some of the lines, just not all of them).

punchagan commented:
Ah, yes! I hadn't thought about that. :) I'll update the commit and let you know.

punchagan commented:
@timabbott I've removed the changes to `lightbox.js` to continue to show previews for Vimeo videos in existing messages.

timabbott commented:
Looks great, merged, thanks @punchagan!
